task/buildah-dance/0.1/buildah-dance.yaml

```diff
@@ -0,0 +1,253 @@
+    script: |
+      #!/bin/python3
+      import json
+
+      # load SBOMs
+      with open("./sbom-image.json") as f:
+        image_sbom = json.load(f)
+
+      with open("./sbom-source.json") as f:
+        source_sbom = json.load(f)
+
+      # fetch unique components from available SBOMs
+      def get_identifier(component):
+        return component["name"] + '@' + component.get("version", "")
+
+      image_sbom_components = image_sbom.get("components", [])
+      existing_components = [get_identifier(component) for component in image_sbom_components]
+
+      source_sbom_components = source_sbom.get("components", [])
+      for component in source_sbom_components:
+        if get_identifier(component) not in existing_components:
+          image_sbom_components.append(component)
+          existing_components.append(get_identifier(component))
+
+      image_sbom_components.sort(key=lambda c: get_identifier(c))
+
+      # write the CycloneDX unified SBOM
+      with open("./sbom-cyclonedx.json", "w") as f:
+        json.dump(image_sbom, f, indent=4)
+    workingDir: $(workspaces.source.path)
+    securityContext:
+      runAsUser: 0
+
+  - name: inject-sbom-and-push
+    image: $(params.BUILDER_IMAGE)
+    computeResources: {}
+    script: |
+      # Expose base image digests
+      buildah images --format '{{ .Name }}:{{ .Tag }}@{{ .Digest }}' | grep -v $IMAGE > $(results.BASE_IMAGES_DIGESTS.path)
+
+      base_image_name=$(buildah inspect --format '{{ index .ImageAnnotations "org.opencontainers.image.base.name"}}' $IMAGE | cut -f1 -d'@')
+      base_image_digest=$(buildah inspect --format '{{ index .ImageAnnotations "org.opencontainers.image.base.digest"}}' $IMAGE)
+      container=$(buildah from --pull-never $IMAGE)
+      buildah config -a org.opencontainers.image.base.name=${base_image_name} -a org.opencontainers.image.base.digest=${base_image_digest} $container
+      buildah commit $container $IMAGE
+
+      status=-1
+      max_run=5
+      sleep_sec=10
+      for run in $(seq 1 $max_run); do
+        status=0
+        [ "$run" -gt 1 ] && sleep $sleep_sec
+        echo "Pushing sbom image to registry"
+        buildah push \
+          --tls-verify=$TLSVERIFY \
+          --digestfile $(workspaces.source.path)/image-digest $IMAGE \
+          docker://$IMAGE && break || status=$?
+      done
+      if [ "$status" -ne 0 ]; then
+          echo "Failed to push sbom image to registry after ${max_run} tries"
+          exit 1
+      fi
+
+      cat "$(workspaces.source.path)"/image-digest | tee $(results.IMAGE_DIGEST.path)
+      echo -n "$IMAGE" | tee $(results.IMAGE_URL.path)
+
+    securityContext:
+      runAsUser: 0
+      capabilities:
+        add:
+          - SETFCAP
+    volumeMounts:
+    - mountPath: /var/lib/containers
+      name: varlibcontainers
+    workingDir: $(workspaces.source.path)
+
+  - name: upload-sbom
```

Comment:
@lcarva, do you have thoughts on whether we should change the process for uploading SBOMs to the registry for customer utilization? Is `cosign attach sbom` still the best option?

Reply:
Funny thing is that `cosign attach sbom` actually uploads the sbom to the registry as a blob which is great. what we do need is to record the digest of the sbom so we can ensure it has not changed post-build. We had this at some point but had to revert because that was the tipping point for exceeding the size of the results in the buildah task.

In konflux, we can get around this because the sbom is always embedded in the image itself, so EC can just use that.

Reply:
We are removing the embedded SBOM for RHTAP in this PR, so we won't be able to rely on that in case we need to. :grimacing: 
(no changes for Konflux though)

Reply:
Oh... Can we introduce the [SBOM_BLOB_URL](https://github.com/redhat-appstudio/build-definitions/pull/654/files) for this task, then?

I'm happy to do so as a follow up PR.

Reply:
I found this is already a requirement for RHTAP (related [comment](https://github.com/redhat-appstudio/build-definitions/pull/729#discussion_r1445274263)). It is supposed to be renamed to `LINK_TO_SBOM`, but I believe it has the same purpose.

Although, I'm confused on how to implement this now. I entered the [deprecate cosign attachment](https://github.com/sigstore/cosign/issues/2755) rabbit hole once I saw the deprecation message in the CLI, and the task currently still is dependent on attach.

Reply:
My suggestion is to keep using the `cosign attach` command as it is done today. It is deprecated, but right now we don't really have a better alternative. The main thing here is that that command uploads the SBOM to the registry as a "blob". So just record the sh256 digest of the SBOM and we should be able to find it in the registry, e.g. `quay.io/foo/bar@sha256:abc..`.

I can help you with this. Feel free to ping me.

Reply:
So, @lcarva, I took the same [approach](https://github.com/redhat-appstudio/build-definitions/pull/729/commits/9912beaddf1ffd58022a90a3069d20aac66d36a2) to generate the digest as in the PR you linked. Thing is, taking the `sha256sum sbom-cyclonedx.json` will give me the digest of the layer, not of the SBOM image that is pushed to the registry. 

You can see the test I did in https://quay.io/repository/bpimente/test-images:
- the SBOM image digest is `ed244f35bd7959a1149836997cd568361dbae6e496ffa560821e0d29c46c94fc`
- the only layer existing has a digest of `57a8c881a24e554cecd66dce96645b55b331138c9613586c3507482b09f0e3c4` 

I can compute the latter if I
```
cosign download sbom quay.io/bpimente/test-images:test-rhtap > sbom.json
sha256sum sbom.json
```

Maybe I'm doing something wrong?

Reply:
I'll move this to a separate PR so we can wrap up the work on this one here.

Reply:
Bruno and I chatted offline, summarizing it here for posterity/transparency. 

We want to record the digest of the layer (which should match the digest of the SBOM file). This is what EC will fetch. An SBOM image could, in theory, contain multiple SBOMs. Adding an SBOM changes the image digest, but not the digest of the other SBOMs (as layers).